File: scripts/refresh-project-prices.py
"""Refresh project price summaries from official historical and current records."""

from __future__ import annotations

import argparse
import csv
import json
import re
import statistics
import urllib.parse
import urllib.request
from collections import defaultdict
from datetime import datetime
from pathlib import Path
from zoneinfo import ZoneInfo


ROOT = Path(__file__).resolve().parents[1]
PROJECTS_PATH = ROOT / "data" / "processed" / "projects.json"
HISTORY_DIR = ROOT / "data" / "source" / "a7-history"
CURRENT_FALLBACK = ROOT / "data" / "source" / "lvr_landcsv" / "h_lvr_land_b.csv"
OFFICIAL_BASE = "https://plvr.land.moi.gov.tw"
OFFICIAL_PAGE = f"{OFFICIAL_BASE}/DownloadOpenData"
FIRST_SEASON = "112S1"


def normalize_name(value: str) -> str:
    return re.sub(r"[\s\-—–・．.（）()第期]", "", value or "").upper()


def roc_to_iso(value: str) -> str | None:
    digits = re.sub(r"\D", "", value or "")
    if len(digits) < 7:
        return None
    year = int(digits[:-4]) + 1911
    return f"{year:04d}-{digits[-4:-2]}-{digits[-2:]}"


def read_csv(path: Path) -> list[dict[str, str]]:
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        rows = list(csv.DictReader(handle))
    return [row for row in rows if row.get("鄉鎮市區") not in {"TOWN", "The villages and towns urban district"}]


def request_bytes(url: str, *, method: str = "GET") -> bytes:
    request = urllib.request.Request(
        url,
        method=method,
        headers={"User-Agent": "Mozilla/5.0 (compatible; JujianDataRefresh/1.0)"},
    )
    with urllib.request.urlopen(request, timeout=60) as response:
        return response.read()


def download_official_sources() -> list[str]:
    HISTORY_DIR.mkdir(parents=True, exist_ok=True)
    season_html = request_bytes(f"{OFFICIAL_BASE}/DownloadSeason_ajax_list", method="POST").decode("utf-8", errors="replace")
    seasons = sorted(set(re.findall(r'<option value="(\d{3}S[1-4])">', season_html)))
    seasons = [season for season in seasons if season >= FIRST_SEASON]
    if not seasons:
        raise RuntimeError("官方歷史季度清單沒有回傳可用資料")

    for season in seasons:
        query = urllib.parse.urlencode({"season": season, "fileName": "H_lvr_land_B.csv"})
        target = HISTORY_DIR / f"{season}_H_lvr_land_B.csv"
        target.write_bytes(request_bytes(f"{OFFICIAL_BASE}/DownloadSeason?{query}"))

    current_query = urllib.parse.urlencode({"fileName": "h_lvr_land_b.csv"})
    (HISTORY_DIR / "current_H_lvr_land_B.csv").write_bytes(
        request_bytes(f"{OFFICIAL_BASE}/Download?{current_query}")
    )
    return seasons


def history_files() -> list[Path]:
    files = sorted(HISTORY_DIR.glob("*_H_lvr_land_B.csv"))
    if not files and CURRENT_FALLBACK.exists():
        files = [CURRENT_FALLBACK]
    if not files:
        raise FileNotFoundError("找不到桃園市預售屋成交來源，請先使用 --download")
    return files


def load_a7_transactions() -> tuple[dict[str, list[dict[str, str]]], int]:
    unique_rows: dict[str, dict[str, str]] = {}
    anonymous_index = 0
    for path in history_files():
        for row in read_csv(path):
            if row.get("鄉鎮市區") != "龜山區" or (row.get("解約情形") or "").strip():
                continue
            serial = (row.get("編號") or "").strip()
            if not serial:
                anonymous_index += 1
                serial = f"anonymous-{path.name}-{anonymous_index}"
            unique_rows[serial] = row

    grouped: dict[str, list[dict[str, str]]] = defaultdict(list)
    for row in unique_rows.values():
        name = normalize_name(row.get("建案名稱", ""))
        try:
            unit_price = float(row.get("單價元平方公尺", ""))
        except (TypeError, ValueError):
            continue
        if name and unit_price > 0:
            grouped[name].append(row)
    return grouped, len(unique_rows)


def street_tokens(value: str) -> list[str]:
    return sorted(set(re.findall(r"[\u4e00-\u9fff0-9一二三四五六七八九十]+(?:路|街|巷)", value or "")))


def summarize(rows: list[dict[str, str]]) -> dict[str, object]:
    prices = sorted(float(row["單價元平方公尺"]) * 3.305785 / 10_000 for row in rows)
    dates = sorted(filter(None, (roc_to_iso(row.get("交易年月日", "")) for row in rows)))
    return {
        "median": round(statistics.median(prices), 1),
        "low": round(prices[0], 1),
        "high": round(prices[-1], 1),
        "count": len(prices),
        "latestDate": dates[-1] if dates else None,
        "source": "內政部預售屋實價登錄（歷史季度＋本期）",
    }


def refresh_projects() -> dict[str, int]:
    payload = json.loads(PROJECTS_PATH.read_text(encoding="utf-8"))
    grouped, reviewed_count = load_a7_transactions()
    checked_at = datetime.now(ZoneInfo("Asia/Taipei")).date().isoformat()
    matched_a7 = 0
    unmatched_a7 = 0

    for project in payload["projects"]:
        if project["region"] == "A7":
            rows = grouped.get(normalize_name(project["name"]), [])
            if rows:
                project["price"] = summarize(rows)
                project["dataCompleteness"] = max(project.get("dataCompleteness", 0), 90)
                tokens = street_tokens(project.get("address", ""))
                corroborated = sorted({token for token in tokens if any(token in row.get("土地位置建物門牌", "") for row in rows)})
                project["priceEvidence"] = {
                    "status": "matched",
                    "statusLabel": "官方成交已配對",
                    "matchMethod": "建案名稱正規化完全相符",
                    "addressCorroborated": bool(corroborated),
                    "addressTokens": corroborated,
                    "lastCheckedAt": checked_at,
                    "sourceUrl": OFFICIAL_PAGE,
                }
                matched_a7 += 1
            else:
                project["price"] = None
                project["dataCompleteness"] = min(project.get("dataCompleteness", 70), 70)
                project["priceEvidence"] = {
                    "status": "official-no-match",
                    "statusLabel": "官方尚無已發布成交",
                    "matchMethod": "官方成交建案名稱未找到可安全歸戶紀錄；地號與地址不足以單獨判定同案",
                    "addressCorroborated": False,
                    "addressTokens": [],
                    "lastCheckedAt": checked_at,
                    "sourceUrl": OFFICIAL_PAGE,
                }
                unmatched_a7 += 1
        else:
            project["priceEvidence"] = {
                "status": "matched" if project.get("price") else "source-no-match",
                "statusLabel": "官方成交已配對" if project.get("price") else "官方來源尚未配對",
                "matchMethod": "新北市開放資料建案名稱配對",
                "addressCorroborated": None,
                "addressTokens": [],
                "lastCheckedAt": checked_at,
                "sourceUrl": "https://data.gov.tw/dataset/146480",
            }

    payload["generatedAt"] = checked_at
    payload["priceCoverage"] = {
        "pricedProjects": sum(1 for project in payload["projects"] if project.get("price")),
        "totalProjects": len(payload["projects"]),
        "a7MatchedProjects": matched_a7,
        "a7OfficialNoMatchProjects": unmatched_a7,
        "a7OfficialRecordsReviewed": reviewed_count,
        "historyFrom": FIRST_SEASON,
        "checkedAt": checked_at,
    }
    for source in payload.get("sources", []):
        if source.get("name") == "內政部預售屋實價登錄":
            source["name"] = "內政部預售屋實價登錄（歷史季度＋本期）"
            source["url"] = OFFICIAL_PAGE
            source["role"] = "A7 自 2023 年起歷史季度與本期成交單價、筆數及最新交易"

    PROJECTS_PATH.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    return {
        "matched_a7": matched_a7,
        "unmatched_a7": unmatched_a7,
        "priced_total": payload["priceCoverage"]["pricedProjects"],
        "reviewed_records": reviewed_count,
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--download", action="store_true", help="先從內政部下載 2023 年起歷史季度與本期資料")
    args = parser.parse_args()
    if args.download:
        seasons = download_official_sources()
        print(f"Downloaded {len(seasons)} official history seasons plus current records")
    result = refresh_projects()
    print(
        f"A7 matched {result['matched_a7']}, official no match {result['unmatched_a7']}; "
        f"total priced {result['priced_total']}/40; reviewed {result['reviewed_records']} official records"
    )


if __name__ == "__main__":
    main()

File: scripts/refresh-project-prices.mjs
import { mkdir, readFile, readdir, writeFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { fetchBytes, normalizeName, parseCsv, rocToIso, todayTaipei } from "./data-utils.mjs";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const projectsPath = path.join(root, "data", "processed", "projects.json");
const a7Directory = path.join(root, "data", "source", "a7-history");
const linkouPath = path.join(root, "data", "source", "linkou_presale_transactions.csv");
const officialBase = "https://plvr.land.moi.gov.tw";
const officialPage = `${officialBase}/DownloadOpenData`;
const linkouDatasetPage = "https://data.ntpc.gov.tw/datasets/D46D75A6-9888-44FF-BBF6-451D5279671A";
const linkouCsvUrl = "https://data.ntpc.gov.tw/api/datasets/D46D75A6-9888-44FF-BBF6-451D5279671A/csv/file";
const firstSeason = "112S1";

function median(numbers) {
  const middle = Math.floor(numbers.length / 2);
  return numbers.length % 2 ? numbers[middle] : (numbers[middle - 1] + numbers[middle]) / 2;
}

function roundOne(value) {
  return Math.round(value * 10) / 10;
}

function streetTokens(value = "") {
  return [...new Set(value.match(/[\p{Script=Han}0-9一二三四五六七八九十]+(?:路|街|巷)/gu) || [])].sort();
}

function summarize(rows, fields, source) {
  const prices = rows
    .map((row) => Number(row[fields.unitPrice]) * 3.305785 / 10_000)
    .filter((value) => Number.isFinite(value) && value > 0)
    .sort((a, b) => a - b);
  const dates = rows.map((row) => rocToIso(row[fields.date])).filter(Boolean).sort();
  return {
    median: roundOne(median(prices)),
    low: roundOne(prices[0]),
    high: roundOne(prices.at(-1)),
    count: prices.length,
    latestDate: dates.at(-1) || null,
    source,
  };
}

async function downloadSources() {
  await mkdir(a7Directory, { recursive: true });
  const seasonBytes = await fetchBytes(`${officialBase}/DownloadSeason_ajax_list`, { method: "POST" });
  const seasonHtml = seasonBytes.toString("utf8");
  const seasons = [...new Set([...seasonHtml.matchAll(/<option value="(\d{3}S[1-4])">/gu)].map((match) => match[1]))]
    .filter((season) => season >= firstSeason)
    .sort();
  if (!seasons.length) throw new Error("官方歷史季度清單沒有回傳可用資料");

  for (const season of seasons) {
    const query = new URLSearchParams({ season, fileName: "H_lvr_land_B.csv" });
    await writeFile(path.join(a7Directory, `${season}_H_lvr_land_B.csv`), await fetchBytes(`${officialBase}/DownloadSeason?${query}`));
  }
  const currentQuery = new URLSearchParams({ fileName: "h_lvr_land_b.csv" });
  await writeFile(path.join(a7Directory, "current_H_lvr_land_B.csv"), await fetchBytes(`${officialBase}/Download?${currentQuery}`));
  await writeFile(linkouPath, await fetchBytes(linkouCsvUrl));
  return seasons.length;
}

async function a7Transactions() {
  const files = (await readdir(a7Directory)).filter((name) => /_H_lvr_land_B\.csv$/iu.test(name)).sort();
  if (!files.length) throw new Error("找不到桃園市預售屋成交來源，請先使用 --download");
  const unique = new Map();
  let anonymous = 0;
  for (const file of files) {
    const rows = parseCsv(await readFile(path.join(a7Directory, file), "utf8"));
    for (const row of rows) {
      if (row.鄉鎮市區 !== "龜山區" || String(row.解約情形 || "").trim()) continue;
      const serial = String(row.編號 || "").trim() || `anonymous-${file}-${anonymous += 1}`;
      unique.set(serial, row);
    }
  }
  const grouped = new Map();
  for (const row of unique.values()) {
    const key = normalizeName(row.建案名稱);
    const unitPrice = Number(row.單價元平方公尺);
    if (!key || !Number.isFinite(unitPrice) || unitPrice <= 0) continue;
    if (!grouped.has(key)) grouped.set(key, []);
    grouped.get(key).push(row);
  }
  return { grouped, reviewedCount: unique.size };
}

async function linkouTransactions() {
  const rows = parseCsv(await readFile(linkouPath, "utf8"));
  const unique = new Map();
  let anonymous = 0;
  for (const row of rows) {
    if (row.district !== "林口區" || String(row.rps30 || "").trim()) continue;
    const serial = String(row.rps27 || "").trim() || `anonymous-${anonymous += 1}`;
    unique.set(serial, row);
  }
  const grouped = new Map();
  for (const row of unique.values()) {
    const key = normalizeName(row.rps28);
    const unitPrice = Number(row.rps22_amountsunitdollars);
    if (!key || !Number.isFinite(unitPrice) || unitPrice <= 0) continue;
    if (!grouped.has(key)) grouped.set(key, []);
    grouped.get(key).push(row);
  }
  return { grouped, reviewedCount: unique.size };
}

function matchedEvidence(project, rows, checkedAt, sourceUrl, addressField, method) {
  const tokens = streetTokens(project.address);
  const corroborated = tokens.filter((token) => rows.some((row) => String(row[addressField] || "").includes(token)));
  return {
    status: "matched",
    statusLabel: "官方成交已配對",
    matchMethod: method,
    addressCorroborated: corroborated.length ? true : null,
    addressTokens: corroborated,
    lastCheckedAt: checkedAt,
    sourceUrl,
  };
}

function unmatchedEvidence(checkedAt, sourceUrl) {
  return {
    status: "official-no-match",
    statusLabel: "官方尚無已發布成交",
    matchMethod: "官方成交建案名稱未找到可安全歸戶紀錄；地號與地址不足以單獨判定同案",
    addressCorroborated: false,
    addressTokens: [],
    lastCheckedAt: checkedAt,
    sourceUrl,
  };
}

async function refreshPrices() {
  const payload = JSON.parse(await readFile(projectsPath, "utf8"));
  const checkedAt = todayTaipei();
  const a7 = await a7Transactions();
  const linkou = await linkouTransactions();
  const regionStats = { A7: { matched: 0, unmatched: 0 }, 林口: { matched: 0, unmatched: 0 } };

  for (const project of payload.projects) {
    const source = project.region === "A7" ? a7 : linkou;
    const rows = source.grouped.get(normalizeName(project.name)) || [];
    const sourceUrl = project.region === "A7" ? officialPage : linkouDatasetPage;
    if (rows.length) {
      project.price = summarize(
        rows,
        project.region === "A7"
          ? { unitPrice: "單價元平方公尺", date: "交易年月日" }
          : { unitPrice: "rps22_amountsunitdollars", date: "rps07" },
        project.region === "A7" ? "內政部預售屋實價登錄（歷史季度＋本期）" : "新北市政府預售屋實價登錄開放資料",
      );
      project.priceEvidence = matchedEvidence(
        project,
        rows,
        checkedAt,
        sourceUrl,
        project.region === "A7" ? "土地位置建物門牌" : "rps02",
        "建案名稱正規化完全相符",
      );
      project.dataCompleteness = Math.max(project.dataCompleteness || 0, 90);
      regionStats[project.region].matched += 1;
    } else {
      project.price = null;
      project.priceEvidence = unmatchedEvidence(checkedAt, sourceUrl);
      project.dataCompleteness = Math.min(project.dataCompleteness || 70, 70);
      regionStats[project.region].unmatched += 1;
    }
  }

  payload.generatedAt = checkedAt;
  payload.priceCoverage = {
    pricedProjects: payload.projects.filter((project) => project.price).length,
    totalProjects: payload.projects.length,
    linkouMatchedProjects: regionStats.林口.matched,
    linkouOfficialNoMatchProjects: regionStats.林口.unmatched,
    linkouOfficialRecordsReviewed: linkou.reviewedCount,
    a7MatchedProjects: regionStats.A7.matched,
    a7OfficialNoMatchProjects: regionStats.A7.unmatched,
    a7OfficialRecordsReviewed: a7.reviewedCount,
    historyFrom: firstSeason,
    checkedAt,
  };
  for (const source of payload.sources || []) {
    if (source.name?.startsWith("內政部預售屋實價登錄")) {
      source.name = "內政部預售屋實價登錄（歷史季度＋本期）";
      source.url = officialPage;
      source.role = "A7 自 2023 年起歷史季度與本期成交單價、筆數及最新交易";
    }
  }
  await writeFile(projectsPath, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
  return { ...payload.priceCoverage };
}

if (process.argv.includes("--download")) {
  console.log(`已下載 ${await downloadSources()} 個歷史季度、本期資料與林口完整成交檔`);
}
const result = await refreshPrices();
console.log(`價格完成：${result.pricedProjects}/${result.totalProjects} 案有官方成交；林口 ${result.linkouMatchedProjects}、A7 ${result.a7MatchedProjects}`);
